dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -360,6 +508,252 @@ std::tuple<RowNumber, bool> WindowTransformAction::stepToFrameEnd(const RowNumbe
+        return std::make_tuple(stepToFrameEndForRangeImpl<UInt64, is_desc>(), true);
+    case TypeIndex::Int8:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Int8, is_desc>(), true);
+    case TypeIndex::Int16:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Int16, is_desc>(), true);
+    case TypeIndex::Int32:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Int32, is_desc>(), true);
+    case TypeIndex::Int64:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Int64, is_desc>(), true);
+    case TypeIndex::Float32:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Float32, is_desc>(), true);
+    case TypeIndex::Float64:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Float64, is_desc>(), true);
+    case TypeIndex::Decimal32:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Decimal32, is_desc>(), true);
+    case TypeIndex::Decimal64:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Decimal64, is_desc>(), true);
+    case TypeIndex::Decimal128:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Decimal128, is_desc>(), true);
+    case TypeIndex::Decimal256:
+        return std::make_tuple(stepToFrameEndForRangeImpl<Decimal256, is_desc>(), true);
+    default:
+        throw Exception("Unexpected column type!");
+    }
+}
+
+template <typename T, bool is_desc>
+RowNumber WindowTransformAction::stepToFrameStartForRangeImpl()
+{
+    return stepToFrameForRangeImpl<T, true, is_desc>();
+}
+
+template <typename T, bool is_desc>
+RowNumber WindowTransformAction::stepToFrameEndForRangeImpl()
+{
+    return stepToFrameForRangeImpl<T, false, is_desc>();
+}
+
+template <typename T, bool is_begin, bool is_desc>
+RowNumber WindowTransformAction::stepToFrameForRangeImpl()
+{
+    RowNumber cursor;
+
+    if constexpr (is_begin)
+        cursor = prev_frame_start;
+    else
+        cursor = prev_frame_end;
+
+    size_t cur_row_aux_col_idx;
+    if constexpr (is_begin)
+        cur_row_aux_col_idx = window_description.frame.begin_range_auxiliary_column_index;
+    else
+        cur_row_aux_col_idx = window_description.frame.end_range_auxiliary_column_index;
+
+    bool is_aux_col_nullable;
+    if constexpr (is_begin)
+        is_aux_col_nullable = window_description.is_aux_begin_col_nullable;
+    else
+        is_aux_col_nullable = window_description.is_aux_end_col_nullable;
+
+    ColumnPtr cur_row_aux_column;
+    if (is_aux_col_nullable)
+    {
+        const auto & nullable_col = static_cast<const ColumnNullable &>(*(inputAt(current_row)[cur_row_aux_col_idx]));
+        cur_row_aux_column = nullable_col.getNestedColumnPtr();
+    }
+    else
+        cur_row_aux_column = inputAt(current_row)[cur_row_aux_col_idx];
+    typename ActualCmpDataType<T>::Type current_row_aux_value = getValue<T>(cur_row_aux_column, current_row.row);
+    return moveCursorAndFindFrameBoundary<typename ActualCmpDataType<T>::Type, is_begin, is_desc>(cursor, current_row_aux_value);
```

Comment:
If it is the function code style of tiflash/clickhouse, the code should be
```
auto column = inputAt(current_row)[cur_row_aux_col_idx];
if (!column->isNullAt(current_row.row))
{
    ...
}
```
Please follow the code style of tiflash.